src/render/filters.rs

```diff
@@ -234,6 +235,70 @@ impl ResolveFilter for ExternalFilter {
     }
 }
 
+impl ResolveFilter for FirstFilter {
+    fn resolve<'t, 'py>(
+        &self,
+        variable: Option<Content<'t, 'py>>,
+        py: Python<'py>,
+        _template: TemplateString<'t>,
+        _context: &mut Context,
+    ) -> ResolveResult<'t, 'py> {
+        let content = match variable {
+            Some(content) => {
+                // Try to get the first item from the content
+                match content {
+                    Content::Py(obj) => {
+                        // Try to get the first item using Python's indexing
+                        // Django only catches IndexError, not TypeError
+                        match obj.get_item(0) {
+                            Ok(first) => Some(Content::Py(first)),
+                            Err(e) => {
+                                // Check if it's an IndexError
+                                if e.is_instance_of::<pyo3::exceptions::PyIndexError>(py) {
+                                    // IndexError -> return empty string
+                                    "".as_content()
+                                } else {
+                                    // Other errors (like TypeError) -> propagate
+                                    return Err(e.into());
+                                }
+                            }
+                        }
+                    }
+                    Content::String(s) => {
+                        // For strings, get the first character
+                        let s_raw = s.as_raw();
+                        if s_raw.is_empty() {
+                            "".as_content()
+                        } else {
+                            // Get the first character
+                            let first_char = s_raw.chars().next().unwrap();
+                            first_char.to_string().into_content()
+                        }
+                    }
+                    Content::Int(_) => {
+                        // Numbers are not sequences, should raise TypeError
+                        // Match Django's behavior exactly
+                        return Err(pyo3::exceptions::PyTypeError::new_err(
+                            "'int' object is not subscriptable",
+                        )
+                        .into());
+                    }
+                    Content::Float(_) => {
+                        // Floats are not sequences, should raise TypeError
+                        // Match Django's behavior exactly
+                        return Err(pyo3::exceptions::PyTypeError::new_err(
+                            "'float' object is not subscriptable",
+                        )
+                        .into());
+                    }
```

Comment:
One of the goals of Django Rusty Templates is to provide better error messages than Django often does, with the help of the `miette` crate. In this case, we could define custom `error::RenderError` variants and then convert them into `PyTypeError` in `template::Template::_render`, like we already do with `VariableDoesNotExist`.